pkg/providers/instanceprofile/instanceprofile.go

```diff
@@ -78,22 +82,18 @@ func (p *DefaultProvider) Get(ctx context.Context, instanceProfileName string) (
 	if err != nil {
 		return nil, err
 	}
-	p.cache.SetDefault(profileCacheKey, out.InstanceProfile)
+	p.instanceProfileCache.SetDefault(profileCacheKey, out.InstanceProfile)
 	return out.InstanceProfile, nil
 }
 
 func (p *DefaultProvider) Create(ctx context.Context, instanceProfileName string, roleName string, tags map[string]string, nodeClassUID string) error {
-	profileCacheKey := getProfileCacheKey(instanceProfileName)
-	roleCacheKey := getRoleCacheKey(roleName)
-	if _, ok := p.cache.Get(roleCacheKey); !ok {
-		out, err := p.iamapi.GetRole(ctx, &iam.GetRoleInput{
-			RoleName: lo.ToPtr(roleName),
-		})
-		if err != nil {
-			return serrors.Wrap(fmt.Errorf("role does not exist, %w", err), "role", roleName)
-		}
-		p.cache.SetDefault(roleCacheKey, out.Role)
+	// Don't attempt to create an instance profile if the role hasn't been found. This prevents runaway instance profile
+	// creation by the NodeClass controller when there's a missing role.
+	if err, ok := p.roleCache.HasError(roleName); ok {
```

Comment:
Probably catching up from the previous round of changes, but why would this cause us to runaway create instance profiles without gating this here?